python/dgl/mock_sparse/diag_matrix.py

```diff
@@ -0,0 +1,261 @@
+        return diag(x, self.shape)
+
+    @property
+    def nnz(self) -> int:
+        """Return the number of non-zero values in the matrix
+
+        Returns
+        -------
+        int
+            The number of non-zero values in the matrix
+        """
+        return self.val.shape[0]
+
+    @property
+    def dtype(self) -> torch.dtype:
+        """Return the data type of the matrix
+
+        Returns
+        -------
+        torch.dtype
+            Data type of the matrix
+        """
+        return self.val.dtype
+
+    @property
+    def device(self) -> torch.device:
+        """Return the device of the matrix
+
+        Returns
+        -------
+        torch.device
+            Device of the matrix
+        """
+        return self.val.device
+
+    def as_sparse(self) -> SparseMatrix:
+        """Convert the diagonal matrix into a sparse matrix object
+
+        Returns
+        -------
+        SparseMatrix
+            The converted sparse matrix object
+
+        Example
+        -------
+
+        >>> import torch
+        >>> val = torch.ones(5)
+        >>> mat = diag(val)
+        >>> sp_mat = mat.as_sparse()
+        >>> print(sp_mat)
+        SparseMatrix(indices=tensor([[0, 1, 2, 3, 4],
+                                     [0, 1, 2, 3, 4]]),
+                     values=tensor([1., 1., 1., 1., 1.]),
+                     shape=(5, 5), nnz=5)
+        """
+        row = col = torch.arange(len(self.val)).to(self.device)
+        return create_from_coo(row=row, col=col, val=self.val, shape=self.shape)
+
+def diag(val: torch.Tensor, shape: Optional[Tuple[int, int]] = None) -> DiagMatrix:
+    """Create a diagonal matrix based on the diagonal values
+
+    Parameters
+    ----------
+    val : torch.Tensor
+        Diagonal of the matrix. It can take shape (N) or (N, D).
+    shape : tuple[int, int], optional
+        If not specified, it will be inferred from :attr:`val`, i.e.,
+        (N, N). Otherwise, :attr:`len(val)` must be equal to :attr:`min(shape)`.
+
+    Returns
+    -------
+    DiagMatrix
+        Diagonal matrix
+
+    Examples
+    --------
+
+    Case1: 5-by-5 diagonal matrix with scaler values on the diagonal
+
+    >>> import torch
+    >>> val = torch.ones(5)
+    >>> mat = diag(val)
+    >>> print(mat)
+    DiagMatrix(val=tensor([1., 1., 1., 1., 1.]),
+               shape=(5, 5))
+    >>> mat.shape
+    (5, 5)
+    >>> mat.nnz
+    5
+
+    Case2: 5-by-10 diagonal matrix with scaler values on the diagonal
+
+    >>> val = torch.ones(5)
+    >>> mat = diag(val, shape=(5, 10))
+    >>> print(mat)
+    DiagMatrix(val=tensor([1., 1., 1., 1., 1.]),
+               shape=(5, 10))
+    >>> mat.shape
+    (5, 10)
+    >>> mat.nnz
+    5
+
+    Case3: 5-by-5 diagonal matrix with tensor values on the diagonal
+
+    >>> val = torch.randn(5, 3)
+    >>> mat = diag(val)
+    >>> mat.shape
+    (5, 5)
+    >>> mat.nnz
+    5
+    """
+    # Mufei: this may not be needed if DiagMatrix is simple enough
```

Comment:
Is this a TODO? or a comment or something else.

Reply:
It's a comment.

Reply:
Then you should use `# NOTE(Mufei): xxx.`

Reply:
Done.